src/lib.rs

```diff
@@ -87,21 +112,42 @@ fn parse_sealed(item: syn::Item) -> syn::Result<proc_macro2::TokenStream> {
 }
 
 // Care for https://gist.github.com/Koxiaet/8c05ebd4e0e9347eb05f265dfb7252e1#procedural-macros-support-renaming-the-crate
-fn parse_sealed_struct(strct: syn::ItemStruct) -> syn::Result<proc_macro2::TokenStream> {
-    let ident = &strct.ident;
+fn parse_sealed_trait(mut item_trait: syn::ItemTrait) -> syn::Result<proc_macro2::TokenStream> {
+    let trait_ident = &item_trait.ident;
+    let trait_generics = &item_trait.generics;
+    let seal = seal_name(trait_ident);
+    item_trait
+        .supertraits
+        .push(parse_quote!(#seal::Sealed #trait_generics));
     Ok(quote!(
-        #strct
-        impl private::Sealed for #ident {}
+        pub(crate) mod #seal {
+            pub trait Sealed #trait_generics {}
+        }
+        #item_trait
     ))
 }
 
-// Care for https://gist.github.com/Koxiaet/8c05ebd4e0e9347eb05f265dfb7252e1#procedural-macros-support-renaming-the-crate
-fn parse_sealed_trait(mut trt: syn::ItemTrait) -> syn::Result<proc_macro2::TokenStream> {
-    trt.supertraits.push(parse_quote!(private::Sealed));
-    Ok(quote!(
-        #trt
-        mod private {
-            pub trait Sealed {}
-        }
-    ))
+fn parse_sealed_impl(item_impl: syn::ItemImpl) -> syn::Result<proc_macro2::TokenStream> {
+    if let Some(impl_trait) = &item_impl.trait_ {
+        let mut sealed_path = impl_trait.1.segments.clone();
+        // since `impl for ...` is not allowed, this path will *always* have at least length 1
+        // thus both `first` and `last` are safe to unwrap
+        let syn::PathSegment { ident, arguments } = sealed_path.pop().unwrap().into_value();
+        let seal = seal_name(ident);
+        sealed_path.push(parse_quote!(#seal));
+        sealed_path.push(parse_quote!(Sealed));
+
+        let self_type = &item_impl.self_ty;
+        let (trait_generics, _, where_clauses) = &item_impl.generics.split_for_impl();
+
+        Ok(quote! {
+            impl #trait_generics #sealed_path #arguments for #self_type #where_clauses {}
+            #item_impl
+        })
+    } else {
+        Err(syn::Error::new_spanned(
+            item_impl,
+            "missing implentation trait",
+        ))
```

Comment:
It's better to early-return this error:

```rust
let impl_trait = item_impl.trait_.as_ref().ok_or_else(|| {
    Err(syn::Error::new_spanned(
        item_impl,
        "missing implentation trait",
    ))
})?;
```

Reply:
Your suggestion does not compile:
```rust
error[E0277]: `?` couldn't convert the error to `syn::Error`
   --> src/lib.rs:117:7
    |
111 | fn parse_sealed_impl(item_impl: syn::ItemImpl) -> syn::Result<TokenStream2> {
    |                                                   ------------------------- expected `syn::Error` because of this
...
117 |     })?;
    |       ^ the trait `From<Result<_, syn::Error>>` is not implemented for `syn::Error`
    |
    = note: the question mark operation (`?`) implicitly performs a conversion on the error value using the `From` trait
    = help: the following implementations were found:
              <syn::Error as From<proc_macro2::LexError>>
    = note: required by `from`
```

Reply:
@jmg-duarte just remove `Err(` there.